app/adapters/abstract_adapter.rb

```diff
@@ -59,6 +59,9 @@ def headers
   JSON_TYPE = Mime[:json]
   private_constant :JSON_TYPE
 
+  DEFAULT_ENCODING = 'utf-8'
+  private_constant :DEFAULT_ENCODING
```

Comment:
I assume you checked `utf-8` is how rails parses the string by default :)

Reply:
I think that strings are converted to UTF8 from whatever they originally were, they are not read as UTF-8 unconditionally. Seems like in the part you link to, they have generated an ASCII-8BIT string that they want now to treat as UTF8 (because it was generated to be valid UTF-8). See

https://github.com/flori/json/blob/v2.6.1/lib/json/pure/parser.rb#L80

So the question is whether `body` is loaded with the supplied `charset=` property or is always treated as ASCII 8bit or is always treated as UTF-8.

My understanding is that if we see charset in the content type to be something, we need to treat that string as the specified charset. It would be nice if rack did that automatically but does it do it?

Reply:
Some more information about the charset.

https://medium.com/@lefloh/when-to-use-a-charset-attribute-for-a-content-type-header-ffeb5838d3c3

json content type is supposed to default to utf-8. In this case I expect things to work with UTF-8. Interesting to know whether Rails automatically uses that information. I'll try to test that.

Reply:
I see that this has nothing to do with Rails. We are using [httpclient](https://github.com/nahi/httpclient) to make requests to the IDP. And it **does** take into account the `charset` parameter. And interprets `body` according to it.

So excuse me for the trouble :bug: 

I think we can just go ahead with your original commit, no need to process the charset by ourselves.

```
$ cat /tmp/index.http 
HTTP/1.1 200 OK
Content-Type: text/html; charset=UTF-16
Content-Length: 7
Server: netcat!

asdfgh

$ nc -l 8000 < /tmp/index.http # in a separate terminal

$ bundle exec httpclient
irb(HTTPClient):023:0> resp = get 'http://localhost:8000'
=> 
#<HTTP::Message:0x00007f296c299388
...
irb(HTTPClient):024:0> resp.content_type
=> "text/html; charset=UTF-16"
irb(HTTPClient):025:0> resp.body
=> "\x61\x73\x64\x66\x67\x68\n"
irb(HTTPClient):026:0> resp.body.encoding
=> #<Encoding:UTF-16 (dummy)>
```

If `charset` is `UTF-8` then httpclient sets `UTF-8` for the body. So I believe we are fine.

And while on it - yes, the body encoding for `application/json` is `UTF-8` if no `charset` is specified.

Reply:
That's all great info @akostadinov Thank you for digging into it and providing the verification steps :clap:

Reply:
Just a note: httpclient implements this feature here https://github.com/nahi/httpclient/commit/e5efea5afb3b5cf6ead3a131644dee71be1ee5e9

ans fixes this issue https://github.com/nahi/httpclient/issues/26